compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -360,19 +393,22 @@ class ExpressionCodegen(
             expression
         )
 
-        val returnType = callee.returnType.substitute(typeSubstitutionMap)
+        val returnType = callee.returnType
         return when {
-            returnType.isNothing() -> {
+            returnType.substitute(typeSubstitutionMap).isNothing() -> {
                 mv.aconst(null)
                 mv.athrow()
                 voidValue
             }
-            expression is IrConstructorCall -> expression.onStack
-            expression is IrDelegatingConstructorCall -> voidValue
+            expression is IrConstructorCall ->
+                MaterialValue(this, asmType, expression.type)
+            expression is IrDelegatingConstructorCall ->
+                voidValue
             expression.type.isUnit() ->
                 // NewInference allows casting `() -> T` to `() -> Unit`. A CHECKCAST here will fail.
-                MaterialValue(mv, callable.returnType).discard().coerce(expression.asmType)
-            else -> MaterialValue(mv, callable.returnType).coerce(expression.asmType)
+                MaterialValue(this, callable.returnType).discard().coerce(expression.type)
```

Comment:
If IrType is nullable because of void, it would probably be best to be consistent and specify it whenever the asm type is non-void, even if it is unused.

Reply:
You're right, I'll make it more consistent.